### PR TITLE
Refactor out-of-time-to-send-messages

### PR DESCRIPTION
**Pull request by peterbarker**
Fixes sensor error loop for Sub and Tracker.

### CODE REVIEW

**Comment by peterbarker** [comment]
It's not clear here, but this changes things so we do care about maximum loop times in Copter and Sub when disarmed.

**Comment by tridge** [comment]
need to be sure that this doesn't impact log download or param download times

**Comment by peterbarker** [comment]
Timings of param-sets and log downloads to be done.

File: libraries/GCS_MAVLink/GCS.h [hunk 1]
```diff
@@ -948,12 +948,14 @@ class GCS
     void update_receive();
     virtual void setup_uarts(AP_SerialManager &serial_manager);
 
-    bool out_of_time() const {
-        return _out_of_time;
-    }
-    void set_out_of_time(bool val) {
-        _out_of_time = val;
+    // maximum amount of time (in microseconds) that must remain in
```
Comment (OXINARF) [inline libraries/GCS_MAVLink/GCS.h]: maximum -> minimum

File: ArduCopter/GCS_Copter.h [hunk 2]
```diff
@@ -32,6 +32,16 @@ class GCS_Copter : public GCS
     bool simple_input_active() const override;
     bool supersimple_input_active() const override;
 
+protected:
+
+    // maximum amount of time (in microseconds) that must remain in
```
Comment (OXINARF) [inline ArduCopter/GCS_Copter.h]: Same.

File: ArduSub/GCS_Sub.h [hunk 3]
```diff
@@ -27,6 +27,16 @@ class GCS_Sub : public GCS
 
     bool vehicle_initialised() const override;
 
+protected:
+
+    // maximum amount of time (in microseconds) that must remain in
```
Comment (OXINARF) [inline ArduSub/GCS_Sub.h]: Same.

File: ArduCopter/GCS_Copter.h (hunk 2 above)
Comment (peterbarker) [reply ArduCopter/GCS_Copter.h]: fixed

File: libraries/GCS_MAVLink/GCS.h (hunk 1 above)
Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS.h]: Fixed.

File: ArduSub/GCS_Sub.h (hunk 3 above)
Comment (peterbarker) [reply ArduSub/GCS_Sub.h]: Fixed.  It was even there in the function name!

**Comment by peterbarker** [comment]
> Is there an actual reason why the minimum amount of time isn't the same across vehicles?

I do not know.

You could probably glean some information from looking at how much work is done in the worst-case scenario when sending a message.  Brace yourself before doing checking that - there's debug in place to trivially show the problems (`GCS_DEBUG_SEND_MESSAGE_TIMINGS`) - because there's some nasty stuff happening in there.

Once this PR is merged notionally we could adaptively change this value based on the currently-measured worst-case value....

**Comment by WickedShell** [comment]
I'd independently recreated this by accident. I'd support just using the tighter 0.25ms window on all platforms, since I'm getting some random misses on quadplanes these days, and its just nicer to reason about. Doesn't have to come in with this now though.

**Comment by peterbarker** [comment]
Copter.

Timings look OK.

```
pr/out-of-time-fixes

Received 962 parameters in 12.37s

STABILIZE> Requested parameter list
Received 962 parameters in 12.37s
Saved 962 parameters to /home/pbarker/rc/flight-data/usb/logs/2019-06-23/flight5/mav.parm
Finished downloading log38.bin (1866712 bytes 8 seconds, 231.2 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 278.4 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 277.1 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 277.7 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 285.6 kbyte/sec 0 retries)


master:
bin/arducopter  962768  1156  195640  1159564
Received 962 parameters in 13.69s
Saved 962 parameters to mav.parm
Received 962 parameters in 12.94s
Saved 962 parameters to mav.parm
Received 962 parameters in 13.09s
Saved 962 parameters to mav.parm
Received 962 parameters in 13.44s
Saved 962 parameters to mav.parm
Received 962 parameters in 13.04s
Saved 962 parameters to mav.parm
Finished downloading log38.bin (1866712 bytes 6 seconds, 271.1 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 290.4 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 290.4 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 276.3 kbyte/sec 0 retries)
Finished downloading log38.bin (1866712 bytes 6 seconds, 280.4 kbyte/sec 0 retries)
```

**Pull request merged**